File: sort/插入排序/insertion.java
/* 
    input:
        4
        1
        5
        2
        3
    output:
        1
        2
        3
        4
        5
*/
public class insertion {
    public static void main(String []args){
        int[] array = {4,1,5,2,3};
        System.out.println("input:");
        for(int a:array)
            System.out.println(a);
        System.out.println("output:");
        for(int b:insertionSort(array))
            System.out.println(b);
    }
    // 插入排序
    public static int[] insertionSort(int []arr){
        for(int i =1;i<arr.length;i++) { 
            // 小于前一个元素
            if (arr[i]<arr[i-1]){
                int index = i-1;
                int temp = arr[i];
                while (index>=0 && arr[index]>temp){
                    arr[index+1] = arr[index];
                    index --;
                }
                arr[index+1] = temp;
            } 
        }
        return arr;
    }
}

File: sort/插入排序/insertion.go
package main

import (
	ozlog "github.com/usthooz/oozlog/go"
)

func main() {
	number := []int{95, 45, 15, 78, 84, 51, 24, 12}
	ozlog.Infof("Sort Result: %v", InsertionSort(number))
}

/*
	InsertionSort:插入排序
	input: [95 45 15 78 84 51 24 12]
	output: [12 15 24 45 51 78 84 95]
*/
func InsertionSort(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		// 小于前一个
		if arr[i] < arr[i-1] {
			index := i - 1
			temp := arr[i]
			// 为元素找到合适的位置
			for index >= 0 && arr[index] > temp {
				arr[index+1] = arr[index]
				index--
			}
			arr[index+1] = temp
		}
	}
	return arr
}
